File: java8/src/test/java/folio/codinginterview/OptimalPortfolioScenarioTest.java
package folio.codinginterview;

import folio.codinginterview.infrastructure.server.DummyServer;
import folio.codinginterview.presentation.PortfolioController;
import org.junit.jupiter.api.Test;

import java.math.BigDecimal;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.assertTrue;

class OptimalPortfolioScenarioTest {

    private static void assertBigDecimalEquals(BigDecimal expected, BigDecimal actual) {
        assertTrue(expected.compareTo(actual) == 0, "expected " + expected + " but was " + actual);
    }

    @Test
    void 最適ポートフォリオを更新取得できる() throws Exception {
        DummyServer server = DummyServer.defaultInstance();
        PortfolioController pc = server.portfolioController();

        // Given: 最適ポートフォリオを Toyopa=0.20, Somy=0.80 に更新する
        pc.updateOptimalPortfolio(new PortfolioController.UpdateOptimalPortfolioRequest(Arrays.asList(
                new PortfolioController.PortfolioItemDto("Toyopa", "0.20"),
                new PortfolioController.PortfolioItemDto("Somy", "0.80")
        ))).get();

        // When: 最適ポートフォリオを取得する
        PortfolioController.GetOptimalPortfolioResponse first = pc.getOptimalPortfolio().get();
        Map<String, String> firstMap = new HashMap<>();
        for (PortfolioController.PortfolioItemDto p : first.portfolios()) {
            firstMap.put(p.symbol(), p.rate());
        }

        // Then: Toyopa=0.20, Somy=0.80 が返される
        assertBigDecimalEquals(new BigDecimal("0.20"), new BigDecimal(firstMap.get("Toyopa")));
        assertBigDecimalEquals(new BigDecimal("0.80"), new BigDecimal(firstMap.get("Somy")));

        // When: 最適ポートフォリオを Toyopa=0.40, Somy=0.60 に更新して再取得する
        pc.updateOptimalPortfolio(new PortfolioController.UpdateOptimalPortfolioRequest(Arrays.asList(
                new PortfolioController.PortfolioItemDto("Toyopa", "0.40"),
                new PortfolioController.PortfolioItemDto("Somy", "0.60")
        ))).get();
        PortfolioController.GetOptimalPortfolioResponse second = pc.getOptimalPortfolio().get();
        Map<String, String> secondMap = new HashMap<>();
        for (PortfolioController.PortfolioItemDto p : second.portfolios()) {
            secondMap.put(p.symbol(), p.rate());
        }

        // Then: Toyopa=0.40, Somy=0.60 が返される
        assertBigDecimalEquals(new BigDecimal("0.40"), new BigDecimal(secondMap.get("Toyopa")));
        assertBigDecimalEquals(new BigDecimal("0.60"), new BigDecimal(secondMap.get("Somy")));
    }
}

File: typescript/tests/optimalPortfolioScenario.test.ts
import Decimal from "decimal.js";
import { describe, it, expect } from "vitest";
import { DummyServer } from "../src/infrastructure/server/dummyServer.js";

describe("Optimal Portfolio Management", () => {
  it("最適ポートフォリオを更新・取得できる", async () => {
    const server = DummyServer.default();
    const pc = server.portfolioController;

    // Given: 最適ポートフォリオを Toyopa=0.20, Somy=0.80 に更新する
    await pc.updateOptimalPortfolio({
      portfolios: [
        { symbol: "Toyopa", rate: "0.20" },
        { symbol: "Somy", rate: "0.80" },
      ],
    });

    // When: 最適ポートフォリオを取得する
    const first = await pc.getOptimalPortfolio();
    const firstMap = new Map(first.portfolios.map((p) => [p.symbol, p.rate]));

    // Then: Toyopa=0.20, Somy=0.80 が返される
    expect(new Decimal(firstMap.get("Toyopa")!).equals(new Decimal("0.20"))).toBe(true);
    expect(new Decimal(firstMap.get("Somy")!).equals(new Decimal("0.80"))).toBe(true);

    // When: 最適ポートフォリオを Toyopa=0.40, Somy=0.60 に更新して再取得する
    await pc.updateOptimalPortfolio({
      portfolios: [
        { symbol: "Toyopa", rate: "0.40" },
        { symbol: "Somy", rate: "0.60" },
      ],
    });
    const second = await pc.getOptimalPortfolio();
    const secondMap = new Map(second.portfolios.map((p) => [p.symbol, p.rate]));

    // Then: Toyopa=0.40, Somy=0.60 が返される
    expect(new Decimal(secondMap.get("Toyopa")!).equals(new Decimal("0.40"))).toBe(true);
    expect(new Decimal(secondMap.get("Somy")!).equals(new Decimal("0.60"))).toBe(true);
  });
});
